x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java

```diff
@@ -797,22 +810,22 @@ LogEntryBuilder withXForwardedFor(ThreadContext threadContext) {
             return this;
         }
 
-        LogEntryBuilder withPrincipal(User user) {
-            logEntry.with(PRINCIPAL_FIELD_NAME, user.principal());
-            if (user.isRunAs()) {
-                logEntry.with(PRINCIPAL_RUN_BY_FIELD_NAME, user.authenticatedUser().principal());
-            }
-            return this;
-        }
-
-        LogEntryBuilder withSubject(Authentication authentication) {
+        LogEntryBuilder withAuthentication(Authentication authentication) {
             logEntry.with(PRINCIPAL_FIELD_NAME, authentication.getUser().principal());
-            if (authentication.getUser().isRunAs()) {
-                logEntry.with(PRINCIPAL_REALM_FIELD_NAME, authentication.getLookedUpBy().getName())
-                        .with(PRINCIPAL_RUN_BY_FIELD_NAME, authentication.getUser().authenticatedUser().principal())
-                        .with(PRINCIPAL_RUN_BY_REALM_FIELD_NAME, authentication.getAuthenticatedBy().getName());
+            logEntry.with(AUTHENTICATION_TYPE_FIELD_NAME, authentication.getAuthenticationType().toString());
+            if (Authentication.AuthenticationType.API_KEY == authentication.getAuthenticationType()) {
+                logEntry.with(API_KEY_ID_FIELD_NAME, (String) authentication.getMetadata().get(ApiKeyService.API_KEY_ID_KEY))
+                        .with(API_KEY_NAME_FIELD_NAME, (String) authentication.getMetadata().get(ApiKeyService.API_KEY_NAME_KEY))
+                        .with(PRINCIPAL_REALM_FIELD_NAME,
+                                (String) authentication.getMetadata().get(ApiKeyService.API_KEY_CREATOR_REALM_NAME));
             } else {
-                logEntry.with(PRINCIPAL_REALM_FIELD_NAME, authentication.getAuthenticatedBy().getName());
+                if (authentication.getUser().isRunAs()) {
+                    logEntry.with(PRINCIPAL_REALM_FIELD_NAME, authentication.getLookedUpBy().getName())
+                            .with(PRINCIPAL_RUN_BY_FIELD_NAME, authentication.getUser().authenticatedUser().principal())
+                            .with(PRINCIPAL_RUN_BY_REALM_FIELD_NAME, authentication.getAuthenticatedBy().getName());
+                } else {
+                    logEntry.with(PRINCIPAL_REALM_FIELD_NAME, authentication.getAuthenticatedBy().getName());
+                }
```

Comment:
I understand these are existing logic and probably not the current focus. But I am troubled by the fact that we have:
* `user.xxx`
* `user.run_by.xxx`
* `user.run_as.xxx`

I personally think we could drop `user.run_by.xxx`. The presence of having `user.run_as.xxx` fields already give the information that `user.xxx` fields are `run_by`. It would be simpler and more consistent imo. Happy to discuss.

Reply:
We can discuss this, but I would prefer we don't change it in this PR.

` user.run_as.*` is only present for `run_as_granted` and `run_as_failed` events. `user.run_by.*` fields are present in all the other events when the user is "run as", so that `user.*` always (run-as or not) conveys the effective user. I think this is consistent.

Reply:
Makes sense to use `user.*` for effective user and `user.run_by.*` when needed. I also agree that we probably do *not* wanna change them for this PR. The same applies to my other suggestion about realm. I do think they need to be tidy up, but may not want sequzee them in for upcoming release.